roscala/src/main/scala/coop/rchain/rosette/prim/Prim.scala

```diff
@@ -18,45 +21,60 @@ abstract class Prim extends Ob {
 
   def fn(ctxt: Ctxt): Either[PrimError, Ob]
 
-  /** Dispatch primitive
-    *
+  /**
     * Rosette seems to potentially return INVALID, UPCALL or DEADTHREAD here
     * Therefore we return RblError for the error case
     */
-  def dispatchHelper(ctxt: Ctxt): Result = {
+  def dispatchHelper: CtxtTransition[Result] = State { ctxt =>
     val n = ctxt.nargs
 
     if (minArgs <= n && n <= maxArgs)
-      fn(ctxt).left.map(PrimErrorWrapper)
+      (ctxt, fn(ctxt).left.map(PrimErrorWrapper))
     else
-      Left(PrimErrorWrapper(mismatchArgs(ctxt, minArgs, maxArgs)))
-  }
-
-  override def dispatch(state: VMState): (Result, VMState) = {
-    // TODO:
-    //if (debugging_level)
-    //    printf("\t%s\n", BASE(id)->asCstring());
-
-    val result = dispatchHelper(state.ctxt)
-
-    if (result.isRight) {
-      val r = result.right.get
-      (Right(r), state.ctxt.ret(r)(state)._2)
-    } else {
-      (result, state)
-    }
-  }
-
-  def invoke(state: VMState): (Result, VMState) = {
-    val result = dispatch(state)
-
-    // TODO:
-    //if (!BASE(ctxt->trgt)->isSynchronousTrgt())
-    //    BASE(ctxt->arg(0))->updateNoArgs();
-
-    result
+      (ctxt, Left(PrimErrorWrapper(mismatchArgs(ctxt, minArgs, maxArgs))))
   }
 
+  /** Dispatch primitive
+    *
+    * This runs a primitive with dispatchHelper and tries
+    * to return the result to the parent ctxt.
+    *
+    * Returning the result to the parent ctxt, will potentially
+    * return the parent ctxt which then has to be scheduled by
+    * the caller.
+    */
+  override def dispatch: CtxtTransition[(Result, Option[Continuation])] =
+    for {
+      primResult <- dispatchHelper
+
+      result <- primResult match {
+                 case Right(ob) =>
```

Comment:
Sorry, could you extract such large expressions into a function? It comes a bit hard to read in `for-comprehensions`.